components/supervisor/pkg/supervisor/git.go

```diff
@@ -103,3 +109,117 @@ func getMissingScopes(required []string, provided map[string]struct{}) []string
 	}
 	return missing
 }
+
+const (
+	minIntervalBetweenGitStatusUpdates = 5 * time.Second
+	lastGitStatusTimeout               = 3 * time.Second
+)
+
+type GitStatusService struct {
+	cfg           *Config
+	content       ContentState
+	git           *git.Client
+	gitpodService serverapi.APIInterface
+	experiments   experiments.Client
+}
+
+type gitStatusUpdateContext struct {
+	lastSuccessfullStatus *gitpod.WorkspaceInstanceRepoStatus
+	statusBackoff         *backoff.ExponentialBackOff
+	updateBackoff         *backoff.ExponentialBackOff
+}
+
+func (s *GitStatusService) Run(ctx context.Context, wg *sync.WaitGroup) {
+	defer wg.Done()
+
+	<-s.content.ContentReady()
+	if ctx.Err() != nil {
+		return
+	}
+
+	updateContext := &gitStatusUpdateContext{
+		statusBackoff: backoff.NewExponentialBackOff(),
+		updateBackoff: backoff.NewExponentialBackOff(),
+	}
+	updateContext.statusBackoff.MaxElapsedTime = 30 * time.Second
+	updateContext.updateBackoff.MaxElapsedTime = 30 * time.Second
+
+	go func() {
+		updates, err := s.gitpodService.WorkspaceUpdates(ctx)
+		if err != nil {
+			log.WithError(err).Error("git: error getting workspace updates")
+			return
+		}
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case update := <-updates:
+				if update == nil {
+					return
+				}
+				updateContext.lastSuccessfullStatus = update.Repo
+			}
+		}
+	}()
+
+	ticker := time.NewTicker(minIntervalBetweenGitStatusUpdates)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			shutdownCtx, shutdown := context.WithTimeout(context.Background(), lastGitStatusTimeout)
+			defer shutdown()
+			updateContext.statusBackoff.Reset()
+			updateContext.updateBackoff.Reset()
+			s.update(shutdownCtx, updateContext)
+			return
+		case <-ticker.C:
+			s.update(ctx, updateContext)
```

Comment:
Would it make sense to use `fsnotify` to watch file changes of git location to get more efficient?

Reply:
And it would be better if we could make sure only one `s.update`` is called (except shutdown update), i.e. use go singleflight.

Since it can be slow due to workspace/cluster state

Reply:
Non-block since I review it too late. They can be improved later if you want 🙈

Reply:
Good ideas, I think it can be improved if we see actual issues during rollout. We are not going to enable it for all users at once.

Reply:
I checked though that in Theia we were pulling as well https://github.com/eclipse-theia/theia/blob/c3fa920a41481c675a3155c2bbb4ebc02bf01b9c/packages/git/src/node/git-repository-watcher.ts#L97C31-L97C31 and it was alright for very big projects too